numpy/core/src/multiarray/buffer.c

```diff
@@ -445,49 +458,22 @@ static PyObject *_buffer_info_cache = NULL;
 static _buffer_info_t*
 _buffer_info_new(PyObject *obj)
 {
+    /*
+     * Note that the buffer info is cached as pyints making them appear like
```

Comment:
Caching as pycapsule objects may speed things up.

Reply:
Plus valgrind would pick it up, that would be nicer.  I doubt it would make real differences for overheads, the creation of a list and dict lookup/insertion should be a much bigger chunk of work.

Reply:
Hashing an arbitrary precision integer should be a lot slower than hashing a pointer.

Reply:
True, I would have to time to judge it. Anyway, moved the pointer increase, as far as I can see that is a trivial move and should be enough.

Reply:
Might be worth trying as a follow-up, obviously not needed in this PR.

May as well be precise here:
```suggestion
     * Note that the buffer info is cached as PyLongObjects making them appear like
```

Reply:
This comment is in a weird place - this function is called _before_ checking the cache. Is that a bug?

Reply:
It was at the perfect place, but now I moved it to the variable definition. The point of putting it there is simple: If you run valgrind with leak-checks it will tell you where the info was allocated, so the comment should be close to the allocation.

It is not a bug, we create the full info, and then just throw it away if its identical (to ensure we do not leak memory if the buffer is exported over and over). It may be a non-ideal design.  (there may be a potential memory leak with respect to relaxed strides checking, which I may fix later.)

I am right now inclined to try the "add a field" path, and see how it goes to be honest...